lib/puma/configuration.rb

```diff
@@ -227,7 +227,12 @@ def puma_default_options(env = ENV)
     def puma_options_from_env(env = ENV)
       min = env['PUMA_MIN_THREADS'] || env['MIN_THREADS']
       max = env['PUMA_MAX_THREADS'] || env['MAX_THREADS']
-      workers = env['WEB_CONCURRENCY']
+      workers = if env['WEB_CONCURRENCY'] == 'auto'
+        require_processor_counter
+        ::Concurrent.available_processor_count
```

Comment:
Hi,

Thanks for this feature it is really handy :)

I was just wondering, is the optimum number the _available_ processors or the _physical_ processors?

I think I'm asking the same question as @MSP-Greg did in this https://github.com/puma/puma/pull/3439#issuecomment-2368500248 🤔 

I tried `WEB_CONCURRENCY=auto` on a 1x vcpu instance on AWS and noticed puma wanted to start up 2x processes.

If that is deliberate and, in your expert opinion, 2 processes per "hyper threaded" (or equivalent) is optimal then all good 😌 

If it isn't then maybe the above can be changed to `::Concurrent.physical_processor_count`.

I would be happy to PR that, but honestly I'm no expert on which would be optimal so thought I'd check first :) 

Here is some console output on the 1vcpu instance showing that `available_processor_count` returns `2.0` while `physical_processor_count` returns `1`.

``` 
> Concurrent.physical_processor_count
1
> ::Concurrent.available_processor_count
2.0
```

Reply:
In my experience, everyone uses logical (available) core count.

Reply:
>I tried WEB_CONCURRENCY=auto on a 1x vcpu instance on AWS and noticed puma wanted to start up 2x processes.

Which instance type is this? I don't think you actually are getting 1 physical/2 logical cores, I think you're on a 2 logical core machine getting 1 logical core worth of CPU time via a hypervisor. vcpu in AWS land is always equivalent to 1 logical core/1 HT.

Reply:
Thanks for the reply 🙏 

It is an aws fargate instance with `cpu` set to `1024`. 
According to the [docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_size) it is one vcpu.

Here is the `lscpu` output for context:

```
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          46 bits physical, 48 bits virtual
  Byte Order:             Little Endian
CPU(s):                   2
  On-line CPU(s) list:    0,1
Vendor ID:                GenuineIntel
  Model name:             Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
    CPU family:           6
    Model:                85
    Thread(s) per core:   2
    Core(s) per socket:   1
    Socket(s):            1
    Stepping:             7
    BogoMIPS:             4999.99
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp
                          lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic mo
                          vbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjus
                          t bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec
                           xgetbv1 xsaves ida arat pku ospke
Virtualization features:
  Hypervisor vendor:      KVM
  Virtualization type:    full
Caches (sum of all):
  L1d:                    32 KiB (1 instance)
  L1i:                    32 KiB (1 instance)
  L2:                     1 MiB (1 instance)
  L3:                     35.8 MiB (1 instance)
```

When I run `top` and press `t` `1` it does show multiple cpus 🤔 

```
top - 02:09:52 up 20:25,  0 user,  load average: 0.02, 0.04, 0.05
Tasks:   7 total,   1 running,   5 sleeping,   0 stopped,   1 zombie
%Cpu0  :   0.7/0.3     1[|                                                                                                   ]
%Cpu1  :   0.7/0.0     1[|                                                                                                   ]
MiB Mem :   3850.9 total,    161.2 free,   1129.5 used,   2842.9 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   2721.3 avail Mem
```

Reads to me like it is actually 1 physical cpu and two logical/ht cores then. Maybe perhaps :)

In which case the current behaviour seems good and might improve performance (we have WEB_CONCURRENCY=1 on production currently on 1vcpu instances that have enough free memory to support two processes).